endpoints/openrtb2/auction.go

```diff
@@ -1931,6 +1965,96 @@ func getAccountID(pub *openrtb2.Publisher) string {
 	return metrics.PublisherUnknown
 }
 
+// Returns the account from the request
+func getAccountIdFromRawRequest(hasStoredRequest bool, storedRequest json.RawMessage, originalRequest []byte) (string, bool, []error) {
+	var accountId string
+	var isAppReq bool
+	var err error
+
+	if hasStoredRequest {
+		accountId, isAppReq, err = searchAccountId(storedRequest)
+		if err != nil {
+			return "", isAppReq, []error{err}
+		}
+	} else {
+		accountId, isAppReq, err = searchAccountId(originalRequest)
+		if err != nil {
+			return "", isAppReq, []error{err}
+		}
+	}
+
+	if accountId == "" {
+		return metrics.PublisherUnknown, isAppReq, nil
+	}
+
+	return accountId, isAppReq, nil
+}
+
+func searchAccountId(request []byte) (accountId string, isAppReq bool, err error) {
+	var exists bool
+	// search in app.publisher.ext.prebid.parentAccount
+	accountId, exists, err = getAccountIdFromPublisherExtInRequest(request, "app")
+	if err != nil {
+		return "", true, err
+	}
+	if !exists {
+		// search in app.publisher.id
+		accountId, exists, err = getAccountIdAsPublisherIdInRequest(request, "app")
+		if err != nil {
+			return "", true, err
+		}
+	}
+	if exists {
+		isAppReq = true
+	}
+	if !exists {
+		// search in site.publisher.ext.prebid.parentAccount
+		accountId, exists, err = getAccountIdFromPublisherExtInRequest(request, "site")
+		if err != nil {
+			return "", false, err
+		}
+	}
+	if !exists {
+		// search in site.publisher.id
+		accountId, exists, err = getAccountIdAsPublisherIdInRequest(request, "site")
+		if err != nil {
+			return "", false, err
+		}
+	}
+
+	return accountId, isAppReq, err
+}
```

Comment:
If I correctly understood the logic, you can simplify it too. Consider suggestion below for the `getAccountIdFromRequest` func.

```suggestion
func searchAccountId(request []byte) (string, bool, error) {
	queries := [...]struct {
		isApp bool
		key  []string
	}{
		{true, []string{"app", "publisher", "ext", openrtb_ext.PrebidExtKey, "parentAccount"}},
		{true, []string{"app", "publisher", "id"}},
		{false, []string{"site", "publisher", "ext", openrtb_ext.PrebidExtKey, "parentAccount"}},
		{false, []string{"site", "publisher", "id"}},
	}

	for _, query := range queries {
		accountId, exists, err := getAccountIdFromRequest(request, query.key)
		if err != nil {
			return "", query.isApp, err
		}

		if exists {
			return accountId, query.isApp, nil
		}
	}
	return "", false, nil
}
```